### ISSUE TITLE
Configuration cache should have a way to exclude known ignorable properties

### ISSUE THREAD

**Issue by ZacSweers** [gradle/gradle#20873]
### Expected Behavior

Currently, configuration cache is sensitive to just about everything. On one hand I understand this, but on the other it's often invalidating on irrelevant things that we don't consume in our build. There should be a way to indicate to gradle that some env vars or properties are _never_ relevant.

### Current Behavior

It will fail on unexpected changes, such as this

```
Calculating task graph as configuration cache cannot be reused because system property 'jnidispatch.path' has changed.
```

I don't know what that property is or why it's changing, we're not using it. There are new APIs for this in 7.5-rc-1, but they pertain to cases where one would intentionally access certain props and vars

> Two new options are now available to mitigate that. For simpler use cases, there are the Provider-based APIs to access [system properties](https://docs.gradle.org/7.5-rc-1/javadoc/org/gradle/api/provider/ProviderFactory.html#systemPropertiesPrefixedBy-java.lang.String-) or [environment variables](https://docs.gradle.org/7.5-rc-1/javadoc//org/gradle/api/provider/ProviderFactory.html#environmentVariablesPrefixedBy-java.lang.String-) with names starting with some prefix. Advanced processing, like filtering names with regular expression, can be done inside a custom [ValueSource](https://docs.gradle.org/7.5-rc-1/javadoc/org/gradle/api/provider/ValueSource.html) implementation. Reading a file, an environment variable, or a system property no longer adds a build configuration input inside the implementation of the ValueSource. The value of the ValueSource is recomputed each time the build runs, and the configuration cache entry is only invalidated if the value changes.

### Context

See above

### Steps to Reproduce 

Unknown, sort of the reason it's a problem

### Your Environment

Build scan URL:

**Comment by ZacSweers** [comment]
Another case where this becomes problematic – we capture system properties about gradle-profiler scenarios/phases for build scans so we can filter by them accordingly, but must forfeit configuration cache in gradle-profiler runs as a result because we cannot indicate they are just for build scans only

```kotlin

// -Dscan.tag.BENCHMARK=ignored
System.getenv("GRADLE_BENCHMARK")?.let {
  tag("BENCHMARK")
  value("BenchmarkScenario", System.getProperty("gradleBenchmarkScenario"))
}
// gradle-profiler injected properties
// Learned from https://github.com/gradle/gradle-profiler/issues/323#issuecomment-788745082
System.getProperty("org.gradle.profiler.phase")?.let {
  value("ProfilerPhase", it)
}
System.getProperty("org.gradle.profiler.number")?.let {
  value("ProfilerNumber", it)
}
System.getProperty("org.gradle.profiler.step")?.let {
  value("ProfilerStep", it)
}
```

**Comment by PaulWoitaschek** [comment]
Another use case:
We use Gitlab CI and inject a CI_JOB_TOKEN that is present for the runners. This is used to authenticate against a private maven repository. 
The effect of this is that on the CI, the config cache is never applied.

**Comment by yogurtearl** [comment]
If you are calling `System.getProperties()` during config or any of your plugins are calling `System.getProperties()` it could trigger this issue and Gradle will see that all system properties where read during config. 

For example `common-custom-user-data-gradle-plugin` before `1.6.3` had a call to `System.getProperties()`

The `System.getProperties()` call was removed here: https://github.com/gradle/common-custom-user-data-gradle-plugin/commit/7d9f4372e9ee99fc85de9aa7bb414afdd1dc00aa#diff-5b3d8f7156853c4ccba4274026d005962fd9ec2e8d92c944867ebf0dc5146f25L88

When you get an unexpected config cache miss due to a system property read, the HTML report will tell you which plugin or script read the system property (if you expand everything).

**Comment by ov7a** [comment]
Sorry for the late reply.

This feature request is in the backlog of the relevant team and is prioritized by them.

**Comment by mlopatkin** [comment]
So far we see some use cases where suppressing an input can be a viable option:
1. Legacy configuration code that cannot be easily changed, but its inputs are benign (e.g. it enumerates system properties to find just one). We're reluctant to provide a blanket option of excluding an input everywhere, but some scope function to call the offending code is not out of question. So far, a `ValueSource` or a `WorkAction` may serve as such a function sometimes, but they have limitations.
2. Tokens and toggles of various sorts that are used at configuration time, but do not really affect the resulting work graph, e.g. a system property that affects configuration-time logging - it is fine to skip the logs when CC entry is reused.
3. Credentials (e.g. short-lived access keys) for dependencies that are needed when resolving the dependencies at store time but do not affect the work graph itself. These are tracked in #24203.
4. Running some work in parallel with the configuration phase, e.g. monitoring thermal throttling.

We are working on providing ways to address these use cases.

There are case where suppressing inputs is not a good solution:
* overzealous build logic written without consideration for inputs in general - this is better to be rewritten with the new APIs Gradle provides.
* tasks that can only be configured eagerly, like setting system properties for the Test task - should be addressed by the migration to Provider API in Gradle 9, and suppressing inputs here is not safe anyway.
* supplying eager values in build scans - `buildscan.background` and `buildscan.buildFinished` can help there.

We also recognize that sometimes it is hard to pinpoint an exact input, especially when it is accessed by a library code. Several improvements are considered here:
* https://github.com/gradle/gradle/issues/30305
* https://github.com/gradle/gradle/issues/23634
* https://github.com/gradle/gradle/issues/26019

**Comment by maxmil** [comment]
We are also keen on this feature. In our case it's the credentials case you mention above. 

In our case it's the GOOGLE_APPLICATION_CREDENTIALS which is a file path containing a token regenerated for each build. This is generated at a different path for each build.

Someway to manually exclude environmental variables from the input sources check would be ideal.

**Comment by steve-todorov** [comment]
@mlopatkin Hey, has there been any new development on this case? This problem is rendering the configuration cache effectively useless for us as it's picking up Github Actions environment variables which obviously change with every build (i.e. `GITHUB_RUN_ID`, `GITHUB_RUN_NUMBER`, `GITHUB_ACTOR_ID`, `GITHUB_SHA`, `GITHUB_STEP_SUMMARY`, etc.

![Image](https://github.com/user-attachments/assets/383f8126-79ea-4806-b76a-87c7f9a817d0)

```
Calculating task graph as configuration cache cannot be reused because environment variable 'GITHUB_STEP_SUMMARY' has changed
```

I think a slightly better approach might be to use an `include` filter instead of `exclude`, because one would know best what env vars the build is actually using or needs.

Another reason to actually make the caching of env vars work with an explicit `include` filter where you define the vars yourself is to prevent leaking credentials in the cache. For example, if your Github Actions workflow is using [Github OIDC with AWS](https://docs.github.com/en/actions/security-for-github-actions/security-hardening-your-deployments/configuring-openid-connect-in-amazon-web-services) you have AWS environment variables available in the workflow. Those might contain a temporary STS AWS Token that would be valid for at least 900s (which is the minimum value you can set for the token). This can be mitigated via using [Gradle Configuration Encryption Key](https://docs.gradle.org/current/userguide/configuration_cache.html#config_cache:secrets:configuring_encryption_key), but if that leaks you could be potentially exposed in a very bad way.

**Comment by trevjonez** [comment]
@steve-todorov one way you might be able to workaround the OIDC issue is by setting up all of your AWS integrations to use profiles pre-stored in `~/.aws` instead of the environment variables. At that point your only input would be the profile name which would/could be static.

```yml
name: OIDC Profile Setup
inputs:
  profile:
    description: name of the profile to configure
    required: true
  role:
    description: ARN for the role to use for the profile
    required: true
  duration:
    description: Length of session in seconds
    required: false
    default: "14400"
  region:
    description: AWS region to use for the profile
    required: false
    default: us-east-1

runs:
  using: "composite"
  steps:
    - name: OIDC Setup - ${{ inputs.profile }} - ${{ inputs.role }}
      shell: bash
      run: |
        mkdir -p ~/.aws
        curl -H "Authorization: bearer $ACTIONS_ID_TOKEN_REQUEST_TOKEN" "$ACTIONS_ID_TOKEN_REQUEST_URL&audience=sts.amazonaws.com" | jq -r '.value' > /tmp/aws-oidc-credentials
        
        aws sts assume-role-with-web-identity \
        --region ${{ inputs.region }} \
        --duration-seconds ${{ inputs.duration }} \
        --role-arn ${{ inputs.role }} \
        --role-session-name GithubActions \
        --web-identity-token file:///tmp/aws-oidc-credentials \
        --query 'Credentials' > /tmp/aws-creds.json
        aws configure set region                "${{ inputs.region }}"                          --profile ${{ inputs.profile }}
        aws configure set aws_access_key_id     "$(jq -r .AccessKeyId     /tmp/aws-creds.json)" --profile ${{ inputs.profile }}
        aws configure set aws_secret_access_key "$(jq -r .SecretAccessKey /tmp/aws-creds.json)" --profile ${{ inputs.profile }}
        aws configure set aws_session_token     "$(jq -r .SessionToken    /tmp/aws-creds.json)" --profile ${{ inputs.profile }}
        
        aws sts get-caller-identity --profile ${{ inputs.profile }}
        
        rm /tmp/aws-oidc-credentials
        rm /tmp/aws-creds.json
```

This also lets you specify a custom duration for the role profile when creating the key pair. Initially I tried just having the profiles in `~/.aws/config` point to the web identity file created by the curl command for their auth, but the web identity seems to always have a 30 minute expiration so longer jobs with aws commands towards the end would fail.

Added bonus is you can setup multiple profiles with different roles so you can retain some level of PoLP when designing your role policies.

**Comment by steve-todorov** [comment]
@trevjonez thanks for sharing your solution! That might solve the AWS problem I mentioned, but unfortunately it won't fix other problems such as the `GITHUB_STEP_SUMMARY` or other changing env vars from [GitHub](https://docs.github.com/en/actions/writing-workflows/choosing-what-your-workflow-does/store-information-in-variables#default-environment-variables). 🫤

**Comment by CLOVIS-AI** [comment]
Another example use-case: when running a Git hook, IntelliJ sets an environment variable `INTELLIJ_SSH_ASKPASS_HANDLER` that has a different value each time. This makes the configuration cache impossible to use within a pre-commit hook, which is a shame because the configuration cache would allow running the build on each commit.

**Comment by marcus-bcl** [comment]
> [@steve-todorov](https://github.com/steve-todorov) one way you might be able to workaround the OIDC issue is by setting up all of your AWS integrations to use profiles pre-stored in `~/.aws` instead of the environment variables. At that point your only input would be the profile name which would/could be static. [...]

@trevjonez I tried to use your workaround for AWS OIDC credentials, but I still get the message:
```
Calculating task graph as configuration cache cannot be reused because file '../../../.aws/credentials' has changed.
```

Any ideas? It feels like we should be able to exclude temporary credentials from the configuration cache invalidation.

For context, I'm using this S3 remote build cache: https://github.com/burrunan/gradle-s3-build-cache

**Comment by steve-todorov** [comment]
Any traction on this case? Now that Gradle 9 is making the config cache "the default", I'd assume this will become higher priority case?

**Comment by ThoSap** [comment]
I also have this issue on GitHub actions and Gradle 9.1
"Calculating task graph as configuration cache cannot be reused because environment variable 'GITHUB_TOKEN' has changed."

In a project, I need to access a Maven Package hosted on GitHub and therefore must pass the `GITHUB_TOKEN` as an environment variable to a private Maven repository.
As the GITHUB_TOKEN changes on every GitHub Action trigger, I can never hit the configuration cache.

As GitHub suggests it.
https://docs.github.com/en/packages/working-with-a-github-packages-registry/working-with-the-gradle-registry#example-using-kotlin-dsl-for-multiple-packages-in-the-same-repository

As I did it, but with no luck.
```kotlin
plugins {
    id("idea")
    id("java")
}

description = "Project"

allprojects {
    group = "com.example"
    version = "0.0.1-SNAPSHOT"

    val githubUser = providers.gradleProperty("gpr.user")
        .orElse(providers.environmentVariable("GITHUB_ACTOR"))
    val githubToken = providers.gradleProperty("gpr.key")
        .orElse(providers.environmentVariable("GITHUB_TOKEN"))

    repositories {
        mavenCentral()
        mavenLocal()

        fun addGithubRepo(name: String) {
            maven {
                this.name = name
                url = uri("https://maven.pkg.github.com/<org_name>/$name")
                credentials {
                    username = githubUser.orNull
                    password = githubToken.orNull
                }
            }
        }

        addGithubRepo("repo1")
        addGithubRepo("repo2")
        addGithubRepo("repo3")
        addGithubRepo("repo4")
    }
}

```

**Comment by ThoSap** [comment]
https://blog.gradle.org/improvements-in-the-build-configuration-input-tracking
> However, certain patterns may lead to a slightly different user experience, requiring a few additional non-cacheable runs before the cached configuration can be used in all subsequent builds. For example, the initial build may create some files after their absence has been recorded, making the cache entry invalid for the next run. Such behavior can be problematic for short-lived environments like ephemeral CI builds.

Due to the ENV var issue, I can never get a configuration cache hit using a GitHub Actions CI, and I won't use a PAT instead of the short-lived token due to security reasons.

**Comment by ThoSap** [comment]
I found issue https://github.com/gradle/gradle/issues/24203

**Comment by ThoSap** [comment]
Related
https://github.com/gradle/gradle/issues/19216
https://github.com/gradle/gradle/issues/22618
https://github.com/gradle/gradle/issues/24203
https://github.com/gradle/gradle/issues/29815

**Comment by woj-tek** [comment]
I just ran into this problem when running the build from CLI on my local machine.
I use iTerm2 & starship.rs for prompt management and each time I'm getting:

```
Calculating task graph as configuration cache cannot be reused because environment variable 'ITERM_PREV_PS1' has changed.
```

Would be nice to be able to exclude certain variables in `gradle.properties`

**Comment by allanconda-mercari** [comment]
My team has been able to successfully enable the configuration cache on CI with outstanding results by leveraging Gradle 8’s handling of the [external processes](https://docs.gradle.org/8.14.3/userguide/configuration_cache.html#config_cache:requirements:external_processes)  limitation, which issues a warning rather than failing the build. Instead of relying on environment variables, we unset the `GITHUB_TOKEN` before running gradle and then retrieve the GitHub Actions token via an external process.

This reduces configuration time on CI by 1–7 minutes and allows tasks to begin execution immediately, fully utilizing available compute resources. This is a significant win for us.

Unfortunately, the same approach now fails with an error instead of warning in Gradle 9, making it a blocker for upgrading.

I strongly encourage reconsidering the priority of supporting this use case. At a minimum, providing a Gradle 9 workaround—such as allowing external processes with a warning rather than a hard failure—would enable teams like ours to upgrade without regressing on CI performance.

**Comment by nuhkoca** [comment]
This is really needed from our side, too. We manage remote cache pushing via environment variable but configuration cache always invalidates when the value changes. However, that environment variable should never affect configuration cache at all. Please make this possible.

**Comment by ed-curran** [comment]
For me the problem was the `io.spring.dependency-management` gradle plugin. It accesses a bunch of system properties. You can see this in the configuration cache report (`build/reports/configuration-cache`). I removed it and was enough for me.
